packages/smart-contracts/test/contracts/ERC20SwapToPay.js

```diff
@@ -0,0 +1,246 @@
+  });
+
+  afterEach(async() => {
+    // The contract should never keep any fund
+    const contractPaymentCcyBalance = await paymentErc20.balanceOf(testSwapToPay.address);
+    const contractRequestCcyBalance = await requestErc20.balanceOf(testSwapToPay.address);
+    expect(contractPaymentCcyBalance.toNumber()).to.equals(0);
+    expect(contractRequestCcyBalance.toNumber()).to.equals(0);
+  })
+
+  expectFromBalanceUnchanged = async() => {
+    const finalFromBalance = await paymentErc20.balanceOf(from);
+    expect(finalFromBalance.toNumber()).to.equals(initialFromBalance.toNumber());
+  }
+
+  it('swaps and pays the request', async function() {
+    let { tx, receipt: { gasUsed } } = await testSwapToPay.swapTransferWithReference(
+      to,
+      10,
+      22,
+      [paymentErc20.address, requestErc20.address],
+      referenceExample,
+      1,
+      builder,
+      exchangeRateOrigin + 100,
+      { from },
+    );
+    await expectEvent.inTransaction(tx, ERC20FeeProxy, 'TransferWithReferenceAndFee', {
+      tokenAddress: requestErc20.address,
+      to,
+      amount: '10',
+      paymentReference: ethers.utils.keccak256(referenceExample),
+      feeAmount: '1',
+      feeAddress: builder,
+    });
+
+    const finalBuilderBalance = await requestErc20.balanceOf(builder);
+    const finalIssuerBalance = await requestErc20.balanceOf(to);
+    expect(finalBuilderBalance.toNumber()).to.equals(1);
+    expect(finalIssuerBalance.toNumber()).to.equals(10);
+    maxGasUsed = gasUsed;
+  });
+
+  it('swaps and pays the request with less gas', async function() {
+    await testSwapToPay.approvePaymentProxyToSpend(requestErc20.address, {
+      from: admin,
+    });
+    await testSwapToPay.approveRouterToSpend(paymentErc20.address, {
+      from: admin,
+    });
+
+    let { tx, receipt: { gasUsed }  } = await testSwapToPay.swapTransferWithReference(
+      to,
+      10,
+      22,
+      [paymentErc20.address, requestErc20.address],
+      referenceExample,
+      1,
+      builder,
+      exchangeRateOrigin + 100,
+      { from },
+    );
+    await expectEvent.inTransaction(tx, ERC20FeeProxy, 'TransferWithReferenceAndFee', {
+      tokenAddress: requestErc20.address,
+      to,
+      amount: '10',
+      paymentReference: ethers.utils.keccak256(referenceExample),
+      feeAmount: '1',
+      feeAddress: builder,
+    });
+
+    const finalBuilderBalance = await requestErc20.balanceOf(builder);
+    const finalIssuerBalance = await requestErc20.balanceOf(to);
+    expect(finalBuilderBalance.toNumber()).to.equals(1);
+    expect(finalIssuerBalance.toNumber()).to.equals(10);
+    // Everything should behave exactly the same, except gas usage.
+    // 50k gas units are saved by approving the router and proxy in advance for first users, per token
+    // console.log(`${maxGasUsed - gasUsed} gas units saved by approving in advance`);
+    expect(gasUsed).to.below(maxGasUsed);
```

Comment:
I wonder of the side effects this could have when running tests... it's usually best to keep each test isolated. maybe not a big deal...

Reply:
How can I make it more isolated?

Reply:
Exchange in DM: not a prio, we should compute "maxGasused" in the same test case.